crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -0,0 +1,599 @@
+
+use crate::{
+    archetype::ArchetypeFlags,
+    query::DebugCheckedUnwrap,
+    system::{EmitEcsEvent, IntoObserverSystem},
+    world::*,
+};
+
+use bevy_ptr::{Ptr, PtrMut};
+use bevy_utils::{EntityHashMap, HashMap};
+
+use crate::{component::ComponentId, prelude::*, world::DeferredWorld};
+
+/// Type used in callbacks registered for observers.
+pub struct Observer<'w, E, B: Bundle = ()> {
+    data: &'w mut E,
+    trigger: ObserverTrigger,
+    _marker: PhantomData<B>,
+}
+
+impl<'w, E, B: Bundle> Observer<'w, E, B> {
+    pub(crate) fn new(data: &'w mut E, trigger: ObserverTrigger) -> Self {
+        Self {
+            data,
+            trigger,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns the event id for the triggering event
+    pub fn event(&self) -> ComponentId {
+        self.trigger.event
+    }
+
+    /// Returns a reference to the data associated with the event that triggered the observer.
+    pub fn data(&self) -> &E {
+        self.data
+    }
+
+    /// Returns a mutable reference to the data associated with the event that triggered the observer.
+    pub fn data_mut(&mut self) -> &mut E {
+        self.data
+    }
+
+    /// Returns a pointer to the data associated with the event that triggered the observer.
+    pub fn data_ptr(&self) -> Ptr {
+        Ptr::from(&self.data)
+    }
+
+    /// Returns the entity that triggered the observer, could be [`Entity::PLACEHOLDER`].
+    pub fn source(&self) -> Entity {
+        self.trigger.source
+    }
+}
+
+#[derive(Default, Clone)]
+pub(crate) struct ObserverDescriptor {
+    events: Vec<ComponentId>,
+    components: Vec<ComponentId>,
+    sources: Vec<Entity>,
+}
+
+/// Metadata for the source triggering an [`Observer`],
+pub struct ObserverTrigger {
+    observer: Entity,
+    event: ComponentId,
+    source: Entity,
+}
+
+// Map between an observer entity and it's runner
+type ObserverMap = EntityHashMap<Entity, ObserverRunner>;
+
+/// Collection of [`ObserverRunner`] for [`Observer`] registered to a particular event targeted at a specific component.
+#[derive(Default, Debug)]
+pub struct CachedComponentObservers {
+    // Observers listening to events targeting this component
+    map: ObserverMap,
+    // Observers listening to events targeting this component on a specific entity
+    entity_map: EntityHashMap<Entity, ObserverMap>,
+}
+
+/// Collection of [`ObserverRunner`] for [`Observer`] registered to a particular event.
+#[derive(Default, Debug)]
+pub struct CachedObservers {
+    // Observers listening for any time this event is fired
+    map: ObserverMap,
+    // Observers listening for this event fired at a specific component
+    component_observers: HashMap<ComponentId, CachedComponentObservers>,
+    // Observers listening for this event fired at a specific entity
+    entity_observers: EntityHashMap<Entity, ObserverMap>,
+}
+
+/// Metadata for observers. Stores a cache mapping event ids to the registered observers.
```

Comment:
This should have a comment explaining how this information can be accessed from a `World`.